### ISSUE TITLE
Have you tested python AELoss and what's the effect?

### ISSUE THREAD

**Issue by FightStone** [Yishun99/nn_utils#1]
Thank you for providing a python AELOSS. Have you tested it? What's the effect? Looking forward to your reply！

**Comment by Yishun99** [comment]
The performance is consistent with the original.

**Comment by FightStone** [comment]
@DouYishun When you test on the whole coco val datasets, not some of them, can you reach the result of paper 62.8 (single) ? I've been struggling with map=57.5(single).

**Comment by Yishun99** [comment]
The result of associative embedding is 56.6 (single scale) reported in paper.
62.8 is achieved by an additional refinement which is same with PAFs.

**Comment by FightStone** [comment]
Yes, you are right, but i doubt i used refinement operation which in pose-ae-train code (map=57.5 single+refinement). Is refinement operation done by this funtion (  refine() which defined in pose-ae-train function.py file  )? Because I don't know much about openpose and refine, I'm getting to know it. looking forward to your reply, thanks very much.

**Comment by Yishun99** [comment]
There is no any "refinement" (as said in associative embedding paper) in pose-ae-train code.

The "refinement" is actually a single-person pose estimator(hourglass) trained on the same dataset.

**Comment by FightStone** [comment]
OK,  I understand, thanks a lot! In other words, "refinement operation" means the middle supervision, is that right? I find pose-ae-train code used it, but in the issue area of AE code, some said Association Embbeding did not meet the paper's targets. Have you reach the result of paper 62.8 (single+refine) ?

**Comment by Yishun99** [comment]
I have not try refinement since it go against the purpose of bottom-up pose estimation.

**Comment by FightStone** [comment]
Thank you for the exquisite manipulation of Tensor in single_image_tag_loss() function.
 If
    return push / ((person_num - 1) * person_num + eps), pull / (person_num + eps)
replacement
    return push * 2 / ((person_num - 1) * person_num + eps), pull / (person_num + eps)
, Would it be better? Maybe the question is foolish.

**Comment by Yishun99** [comment]
Sorry for the delay, I did not find any difference expect the "* 2".

**Comment by FightStone** [comment]
"*2" means balancing the losses of “push” and “pull” (e.g magnitude). In addition, replace hourglass with hrnet, i reach the result of map=61.3 (single scale and without refinement) in cocodev with some trick, but it still far away with map=65.0, can you give me some advices to further improve. Thx!

**Comment by Yishun99** [comment]
Try multi-scale test. Almost all of the method in MPII use multi-scale test. And AE gain 8.4 mAP from multi-scale test reported in paper.
